packages/base-driver/lib/basedriver/driver.js

```diff
@@ -382,116 +115,267 @@ class BaseDriver extends Protocol {
-    }
+    this.noCommandTimer = setTimeout(async () => {
+      this.log.warn(
+        `Shutting down because we waited ` +
+          `${this.newCommandTimeoutMs / 1000.0} seconds for a command`,
+      );
+      const errorMessage =
+        `New Command Timeout of ` +
+        `${this.newCommandTimeoutMs / 1000.0} seconds ` +
+        `expired. Try customizing the timeout using the ` +
+        `'newCommandTimeout' desired capability`;
+      await this.startUnexpectedShutdown(new Error(errorMessage));
+    }, this.newCommandTimeoutMs);
   }
 
-  /*
-   * Restart the session with the original caps,
-   * preserving the timeout config.
+  /**
+   *
+   * @param {import('@appium/types').AppiumServer} server
+   * @param {string} host
+   * @param {number} port
+   * @param {string} path
    */
+  assignServer (server, host, port, path) {
+    this.server = server;
+    this.serverHost = host;
+    this.serverPort = port;
+    this.serverPath = path;
+  }
+
+  /*
+    * Restart the session with the original caps,
+    * preserving the timeout config.
+    */
   async reset () {
     this.log.debug('Resetting app mid-session');
     this.log.debug('Running generic full reset');
 
     // preserving state
     let currentConfig = {};
-    for (let property of ['implicitWaitMs', 'newCommandTimeoutMs', 'sessionId', 'resetOnUnexpectedShutdown']) {
+    for (let property of [
+      'implicitWaitMs',
+      'newCommandTimeoutMs',
+      'sessionId',
+      'resetOnUnexpectedShutdown',
+    ]) {
       currentConfig[property] = this[property];
     }
 
     // We also need to preserve the unexpected shutdown, and make sure it is not cancelled during reset.
     this.resetOnUnexpectedShutdown = () => {};
 
     try {
-      await this.deleteSession(this.sessionId);
+      if (this.sessionId !== null) {
+        await this.deleteSession(this.sessionId);
+      }
       this.log.debug('Restarting app');
-      await this.createSession(undefined, undefined, this.originalCaps);
+      await this.createSession(this.originalCaps);
     } finally {
       // always restore state.
       for (let [key, value] of _.toPairs(currentConfig)) {
         this[key] = value;
       }
     }
-    this.clearNewCommandTimeout();
+    await this.clearNewCommandTimeout();
   }
 
-  proxyActive (/* sessionId */) {
-    return false;
-  }
+  /**
+     *
+     * Historically the first two arguments were reserved for JSONWP capabilities.
+     * Appium 2 has dropped the support of these, so now we only accept capability
+     * objects in W3C format and thus allow any of the three arguments to represent
+     * the latter.
+     * @param {W3CCapabilities} w3cCapabilities1
+     * @param {W3CCapabilities} [w3cCapabilities2]
+     * @param {W3CCapabilities} [w3cCapabilities]
+     * @param {DriverData[]} [driverData]
+     * @returns {Promise<[string,object]>}
+     */
+  async createSession (
+    w3cCapabilities1,
+    w3cCapabilities2,
+    w3cCapabilities,
```

Comment:
is it a bit strange this one doesn't have a number? or that if it doesn't, that it's not first in the list?